File: _experimental/_c++/words/test/words_test.cpp
#include <gtest/gtest.h>
#include "../include/words.h"

class WordsTest : public ::testing::Test
{
protected:
    Words _words;
};

TEST_F(WordsTest, Reverse_String)
{
    EXPECT_EQ(_words.reverse_string("hello"), "olleh");
    EXPECT_EQ(_words.reverse_string("a"), "a");
    EXPECT_EQ(_words.reverse_string(""), "");
}

TEST_F(WordsTest, Is_Palindrome)
{
    EXPECT_EQ(_words.is_palindrome("race car"), true);
    EXPECT_EQ(_words.is_palindrome("level"), true);
    EXPECT_EQ(_words.is_palindrome("hello"), false);
    EXPECT_EQ(_words.is_palindrome("a"), true);
    EXPECT_EQ(_words.is_palindrome(""), true);
}

TEST_F(WordsTest, Is_Substring)
{
    EXPECT_EQ(_words.kmp_search("test", "this is a test"), true);
    EXPECT_EQ(_words.kmp_search("not", "this is a test"), false);
    EXPECT_EQ(_words.kmp_search("", "any string"), true);
    EXPECT_EQ(_words.kmp_search("abc", "abc"), true);
    EXPECT_EQ(_words.kmp_search("abc", "ab"), false);
}

TEST_F(WordsTest, Longest_Common_Subsequence)
{
    EXPECT_EQ(_words.lcs("AGGTAB", "GXTXAYB"), 4);
    EXPECT_EQ(_words.lcs("ABC", "ABC"), 3);
    EXPECT_EQ(_words.lcs("ABC", "DEF"), 0);
    EXPECT_EQ(_words.lcs("", "ABC"), 0);
    EXPECT_EQ(_words.lcs("ABC", ""), 0);
}

File: _experimental/_erlang/words/test/words_test.erl
-module(words_test).
-include_lib("eunit/include/eunit.hrl").

reverse_string_test() ->
    ?assertEqual("olleh", words:reverse_string("hello")),
    ?assertEqual("a", words:reverse_string("a")),
    ?assertEqual("", words:reverse_string("")).

is_palindrome_test() ->
    ?assertEqual(true, words:is_palindrome("race car")),
    ?assertEqual(true, words:is_palindrome("level")),
    ?assertEqual(false, words:is_palindrome("hello")),
    ?assertEqual(true, words:is_palindrome("a")),
    ?assertEqual(true, words:is_palindrome("")).

is_substring_test() ->
    ?assertEqual(true, words:kmp_search("test", "this is a test")),
    ?assertEqual(false, words:kmp_search("not", "this is a test")),
    ?assertEqual(true, words:kmp_search("", "any string")),
    ?assertEqual(true, words:kmp_search("abc", "abc")),
    ?assertEqual(false, words:kmp_search("abc", "ab")).

largest_common_subsequence_test() ->
    ?assertEqual(4, words:lcs("AGGTAB", "GXTXAYB")),
    ?assertEqual(3, words:lcs("ABC", "ABC")),
    ?assertEqual(0, words:lcs("ABC", "DEF")),
    ?assertEqual(0, words:lcs("", "ABC")),
    ?assertEqual(0, words:lcs("ABC", "")).